primitives/src/sync.rs

```diff
@@ -0,0 +1,149 @@
+use core::ops::{Deref, DerefMut};
+use std::{
+    sync::atomic::{AtomicU32, Ordering},
+    time::{Duration, Instant},
+};
+
+const TIME_SPINNING_MICRO_SECONDS: u128 = 500;
+const SLEEP_TIME: Duration = Duration::from_micros(100);
+
+pub struct MutexMutGuard<'a, T: ?Sized> {
+    mutex: &'a Mutex<T>,
+}
+
+pub struct MutexGuard<'a, T: ?Sized> {
+    mutex: &'a Mutex<T>,
+}
+
+pub struct Mutex<T: ?Sized> {
+    counter: AtomicU32,
+    interior: T,
+}
+
+impl<T: ?Sized> std::fmt::Debug for Mutex<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "iroha_mutex<{}>", std::any::type_name::<T>())
+    }
+}
+
+impl<T> Mutex<T> {
+    pub fn new(interior: T) -> Self {
+        Self {
+            counter: AtomicU32::new(0),
+            interior,
+        }
+    }
+
+    pub fn lock_mut(&self) -> MutexMutGuard<T> {
+        let instant_entered = Instant::now();
+        loop {
+            match AtomicU32::compare_exchange_weak(
+                &self.counter,
+                0,
+                1,
+                Ordering::AcqRel,
+                Ordering::Relaxed,
+            ) {
+                Ok(_) => break,
+                Err(prev) => debug_assert!(prev == 1),
+            }
+            if instant_entered.elapsed().as_micros() > TIME_SPINNING_MICRO_SECONDS {
+                std::thread::sleep(SLEEP_TIME);
+            }
+        }
+        MutexMutGuard::<T> { mutex: &self }
+    }
+
+    pub fn lock(&self) -> MutexGuard<T> {
+        let instant_entered = Instant::now();
+        loop {
+            match AtomicU32::compare_exchange_weak(
+                &self.counter,
+                0,
+                1,
+                Ordering::AcqRel,
+                Ordering::Relaxed,
+            ) {
+                Ok(_) => break,
+                Err(prev) => debug_assert!(prev == 1),
+            }
+            if instant_entered.elapsed().as_micros() > TIME_SPINNING_MICRO_SECONDS {
+                std::thread::sleep(SLEEP_TIME);
+            }
+        }
```

Comment:
`compare_exchange_weak` can fail spuriously. Therefore, I believe that the `prev` might also be `0` and the assertion `prev == 1` you make is incorrect. 
If this is true then I suggest the following:

```suggestion
        while let Err(_) = self.counte.compare_exchange_weak(
                0, 1, Ordering::AcqRel, Ordering::Relaxed,
        ) {
            if instant_entered.elapsed().as_micros() > TIME_SPINNING_MICRO_SECONDS {
                std::thread::sleep(SLEEP_TIME);
            }
        }
```